keanu-project/src/main/java/io/improbable/keanu/network/LambdaSection.java

```diff
@@ -121,4 +143,57 @@ public static LambdaSection getDownstreamLambdaSection(Vertex<?> aVertex, boolea
 
         return result;
     }
+
+    /**
+     * @param vertices       vertices to start propagation from
+     * @param nextVertices The next vertices to move to given a current vertex. E.g getChildren for downstream or
+     *                     getParents for upstream.
+     * @param shouldAdd    true when a given vertex should be included in the result, false otherwise
+     * @return A Set of vertices that are in the direction implied by nextVertices and filtered by shouldAdd
+     */
+    public static Set<Vertex> getVerticesCollection(List<Vertex> vertices,
+                                                    Function<Vertex, Collection<Vertex>> nextVertices,
+                                                    Predicate<Vertex> shouldAdd) {
```

Comment:
there's a lot of duplication between this and `getVertices` - I expect it's quite easy to roll them into one method

Reply:
... and once you've done that, it should be easy to remove the duplication between  `getUpstreamLambdaSectionForCollection` and `getUpstreamLambdaSection` too